projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs

```diff
@@ -961,8 +960,32 @@ private enum RecoveryConnectionState
         private Task _recoveryTask;
         private RecoveryConnectionState _recoveryLoopState = RecoveryConnectionState.Connected;
 
-        private readonly ConcurrentQueue<RecoveryCommand> _recoveryLoopCommandQueue = new ConcurrentQueue<RecoveryCommand>();
-        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+
+        private class AsyncConcurrentQueue<T>
+        {
+            private readonly ConcurrentQueue<T> _internalQueue = new ConcurrentQueue<T>();
+            private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+
+            public async Task<T> DequeueAsync(CancellationToken token)
+            {
+                await _semaphore.WaitAsync(token).ConfigureAwait(false);
+
+                if (!_internalQueue.TryDequeue(out var command))
+                {
+                    throw new InvalidOperationException("Internal queue empty despite signaled enqueue.");
+                }
+
+                return command;
+            }
+
+            public void Enqueue(T item)
+            {
+                _internalQueue.Enqueue(item);
+                _semaphore.Release();
+            }
+        }
```

Comment:
Move this to a separate file in the Utils namespace.

Reply:
Done